Refactor of blockchain's accounts

jrainville opened the pull request:
This is the first step of implementing the new accounts for the blockchain config (as badly described here: https://github.com/status-im/embark-specs/issues/13)

Since the spec is not super clear, let me try to explain what this PR does:
- Get rid of `account{}` and `simulatorMnemonic` as they were confusing
 - The mnemonic used by the simulator now comes from `accounts[x].mnemonic` (so if there's a mnemonic in the accounts array)
- Add the same account configuration as in the contract config (`accounts` array with multiple types of accounts)
 - Accounts using `nodeAccount: true` count as "real" node account (so geth and parity accounts that need to be created and unlocked)
   - Those use the password field to create the account and unlock it
   - In the spec, there was `promptPassword`, but I didn't add it yet, because it's way more complex and this PR is already pretty big
 - All other account types are kind of like "proxy" accounts. They count as part of the `nodeAccounts` in the contract config (so they are used as if they came directly from the node), but are not unlocked by Geth or Parity.

With the way I did it right now, it would be super easy to actually get rid of the accounts in contracts config and only use the ones in blockchain, as they use the same params and are both mostly processed by the provider.

I tested this using Geth, Parity and the simulator, but it still is a breaking change, so would need more tests to be really sure it's not bugged.

![4jyjurz](https://user-images.githubusercontent.com/11926403/48952449-30150800-ef10-11e8-8882-6dd28db87f2d.png)

jrainville commented:
Remind me to write docs for this

alaibe commented on src/lib/modules/blockchain_process/blockchain.js:
```diff
@@ -27,6 +28,13 @@ var Blockchain = function(userConfig, clientClass) {
   this.proxyIpc = null;
   this.isStandalone = userConfig.isStandalone;
 
+  if (this.userConfig.account) {
```
Should this be in the config parser?
Arriving at this point I would expect the config to be valid

alaibe commented on src/lib/modules/blockchain_process/blockchain.js:
```diff
@@ -37,7 +45,7 @@ var Blockchain = function(userConfig, clientClass) {
     networkId: this.userConfig.networkId || clientClass.DEFAULTS.NETWORK_ID,
     genesisBlock: this.userConfig.genesisBlock || false,
-    datadir: this.userConfig.datadir || false,
+    datadir: this.userConfig.datadir || '.embark/' + this.env + '/datadir',
```
can default dir be set in default of the config?

jrainville commented:
@alaibe Changes done. It's actually better like you asked for the validation, because it fails quicker that way.
Tell me when it's good so that I can squash again.

jrainville commented:
This is ready for a final review. Added the docs urls and rebased.

Might be worth to wait for the next alpha release before merging though, to be sure.

PascalPrecht commented on src/lib/core/config.js:
```diff
@@ -212,11 +212,13 @@ Config.prototype._getFileOrOject = function(object, filePath, property) {
-      "proxy": true
+    default: {
+      enabled: true,
+      ethereumClientName: 'geth',
```
Can we put `'get'` in a constant?

PascalPrecht commented on src/lib/modules/blockchain_process/blockchain.js:
```diff
@@ -63,32 +64,42 @@ var Blockchain = function(userConfig, clientClass) {
+    }
+  }
+
+  if (this.userConfig === {} || this.userConfig.default || JSON.stringify(this.userConfig) === '{"ethereumClientName":"geth"}') {
```
```
JSON.stringify(this.userConfig) === '{"ethereumClientName":"geth"}'
```

This looks very weird to me. Any reason we don't check for 

```
userConfig.ethereumClientName && userConfig.ethereumClientName === 'geth'
```

jrainville replied:
From what I understand, stringify removes all the falsy values, so we end up with just `'{"ethereumClientName":"geth"}'`. Only testing for the `ethereumClientName` would be bad, because we would override to the defaults pretty much all the time, whereas this checks that we used no configs.
Anyway, I agree it's weird. That use case it pretty much only used by the unit tests from what I can see.

PascalPrecht commented on src/lib/modules/blockchain_process/blockchain.js:
```diff
@@ -402,15 +406,15 @@ Blockchain.prototype.initChainAndGetAddress = function (callback) {
           return next();
         }
-        this.logger.info(__("already initialized").green);
+        self.logger.info(__("already initialized").green);
```
I'd **actually** prefer if we'd move from `self` to `this` and use fat arrow functions. I assume this has been changed for consistency reasons?

jrainville replied:
I changed it because it crashed yes haha

PascalPrecht commented on src/lib/utils/accountParser.js:
```diff
@@ -44,8 +44,11 @@ class AccountParser {
+        if (typeof account === 'string') {
+          return {address: account};
+        }
+        return account;
```
Sugar, but how about

```
return nodeAccounts.map(account => typeof account === 'string' ? { address: account } : account)
```

PascalPrecht commented on src/test/config.js:
```diff
@@ -41,6 +44,7 @@ describe('embark.Config', function () {
     it('should convert Ether units', function () {
       let expectedConfig = {
         "enabled": true,
+        "ethereumClientName": "geth",
```
It would be really cool if we could just do

```
import { GETH_CLIENT_NAME } from '@embark/core';
```

And use that. If we make `geth` a constant somewhere in our source, we can at least require it from there.

alaibe commented:
> From what I can see in the changes, it doesn't seem to be too hard to keep support for the older syntax, which would mean we wouldn't break people's apps.

I would disagree with this statement, it is super hard in the current state.
If you look at the config, this is extremely complex, sometimes we validate in config, sometimes in the module, sometimes not, sometimes there is default, sometimes not...
The inputs are not even validated

I could carry on like that a lot.
So that means that if this is weak, it is easily broken, it is complex to understand for the end user.
Hence why I would keep to the minimum until we have an implementation that allows us to be more friendly.

PascalPrecht commented:
Okay, fair enough. I was just hoping there'd be a chance to not introduce a breaking change if possible. I probably underestimate the current state of affairs.

But I fully agree, from a user perspective it should just be as easy as possible.

jrainville commented:
So for context, we wanted to remove because it was confusing to have different types of accounts between blockchain and contracts. Also, that way, we can slowly deprecate contract accounts and only use the blockchain accounts. Here is the spec we created back in the days: https://github.com/status-im/embark-specs/issues/13

We could support both, but if `accounts` (new) is specified, we ignore the old one, which can be confusing (even more) for the user. So I think that breaking now that we move to a major might be ok. What do you think?

PascalPrecht commented:
> Also, that way, we can slowly deprecate contract accounts and only use the blockchain accounts.

That actually answers my question in: https://github.com/embark-framework/embark-site/pull/113#discussion_r237005382

So blockchain accounts will replace contracts accounts?

> So I think that breaking now that we move to a major might be ok. 

Yea, we're definitely in a "good" position to place a breaking change. In general, however, I'd always try to push those back as much as we can. 

> We could support both, but if accounts (new) is specified, we ignore the old one, which can be confusing (even more) for the user

This is a matter of documenting those clearly. In fact, we have already several spots in the docs where we focus on the newer syntaxes and then add info boxes about older (but still available) syntaxes.

Anyways, if it's indeed too complex I'm fine with introducing a breaking change considering that we're on a major release alpha phase.

jrainville commented:
Just checked with Iuri and I have a small little improvement to make before merging.
